apps/editing-toolkit/editing-toolkit-plugin/tutorials/class-wpcom-tutorials.php

```diff
@@ -0,0 +1,227 @@
+<?php
+/**
+ * Tutorials (We would have called them Guided Tours but it's been done).
+ *
+ * Each Tutorial contains a flat list of Tasks. Each Task may contain several Steps in
+ * a `WpcomTourKit` component on the frontend, but we are only concerned with the top
+ * two levels of Tutorial and its contained Tasks, tracking Task completion status
+ * on the backend. This API will be used to generate
+ *
+ * @package A8C\FSE
+ */
+
+/**
+ * WPCom_Tutorials class
+ */
+class WPCom_Tutorials {
+
+	/**
+	 * Storage for our singleton instance.
+	 *
+	 * @var null|WPCom_Tutorials
+	 */
+	private static $instance = null;
+
+	/**
+	 * Internal storage for registered Tutorials.
+	 *
+	 * @var array
+	 */
+	private $registry = array();
+
+	/**
+	 * Gets aour singleton instance, initiating it the first time.
+	 *
+	 * @return WPCom_Tutorials instance.
+	 */
+	public static function get_instance() {
+		if ( is_null( self::$instance ) ) {
+			self::$instance = new self();
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Registers a Tutorial.
+	 *
+	 * @param string $id      A unique ID for the Tutorial.
+	 * @param array  $options `title` and `tasks` required.
+	 *
+	 * @return bool           True if successfully registered, false if it failed.
+	 */
+	public function register( $id, $options ) {
+		// @todo verify tasks strutcture
+		if ( ! $this->validate_options( $options ) ) {
+			return false;
+		}
+		$options['id']         = $id;
+		$this->registry[ $id ] = $options;
+		return true;
+	}
+
+	/**
+	 * Unregisters a registered Tutorial
+	 *
+	 * @param string $id Tutorial ID.
+	 * @return bool      True if successfully unregistered, false if not found.
+	 */
+	public function unregister( $id ) {
+		if ( ! isset( $this->registry[ $id ] ) ) {
+			return false;
+		}
+		unset( $this->registry[ $id ] );
+		return true;
+	}
+
+	/**
+	 * Validates the options passed in `WPCom_Tutorials::register`.
+	 *
+	 * @param array $options Associative array passed with registration.
+	 * @return bool          Validation success.
+	 */
+	private function validate_options( $options ) {
+		if ( ! isset( $options['title'] ) || ! is_string( $options['title'] ) ) {
+			return false;
+		}
+		if ( ! isset( $options['tasks'] ) || ! is_array( $options['tasks'] ) ) {
+			return false;
+		}
+		foreach ( $options['tasks'] as $task ) {
+			if ( ! $this->validate_task( $task ) ) {
+				return false;
+			}
+		}
+		return true;
+	}
+
+	/**
+	 * Validates a single task for expected structure.
+	 *
+	 * @param array $task A single task (id and title).
+	 * @return bool       Validation success.
+	 */
+	private function validate_task( $task ) {
+		return isset( $task['id'] ) && isset( $task['title'] );
```

Comment:
should we also validate types here similar to what we do in `validate_options`?

Reply:
If I'm understanding Addie correctly, we call `is_string` for options title and `is_array` for options tasks in the `validate_options` method. Do we need to call the appropriate type validations for task id and task title in validate_task?

Reply:
> We call is_string for options title and is_array for options tasks in the validate_options method. Do we need to call the appropriate type validations for task id and task title in the validate_task?

Thats what I was getting at 😁 . I noticed we were calling this is_type sort of functions for validate_options but not validate_tasks.

Reply:
Thanks for sorting me out @jeyip I sure completely misread @Addison-Stavlo's comment.

Now that I see the point properly, yes, strictly speaking, those could also be validated more, um, strictly. On a practical level I struggle to envision somebody crafting a non-string payload for those for any reason whatsoever, to any profit whatsoever, but people, being what they are, one should not make assumptions: https://github.com/Automattic/wp-calypso/pull/64510/commits/8e2c01e7a3fa171264c25c48c35af0893f336a90